dandiapi/api/views/asset.py

```diff
@@ -72,6 +75,47 @@ def _download_asset(asset: Asset):
         raise ValueError(f'Unknown storage {storage}')
 
 
+def _paginate_asset_paths(
+    items: list, page: int, page_size: int, versions__dandiset__pk: str, versions__version: str
+) -> dict:
+    """Paginates a list of files and folders to be returned by the asset paths endpoint."""
+    if page_size > DandiPagination.max_page_size:
+        page_size = DandiPagination.max_page_size
+
+    asset_count: int = len(items)
+
+    paths_paginator: Paginator = Paginator(list(items.items()), page_size)
+    assets_page: Page = paths_paginator.page(page)
+
+    # Divide into folders and files
+    paths = {'folders': {}, 'files': {}}
+    folder_files = [paths['folders'], paths['files']]
+
+    # Note that this loop runs in constant time, since the length of assets_page
+    # will never exceed DandiPagination.max_page_size.
+    for k, v in dict(assets_page).items():
+        folder_files[int(isinstance(v, Asset))][k] = v
+
+    paths = AssetPathsSerializer(paths)
+
+    # generate other parameters for the paginated response
+    url_kwargs = {
+        'versions__dandiset__pk': versions__dandiset__pk,
+        'versions__version': versions__version,
+    }
+    url = reverse('asset-paths', kwargs=url_kwargs)
+    next_page = page + 1 if page + 1 <= paths_paginator.num_pages else None
+    prev_page = page - 1 if page - 1 > 0 else None
+    next_url = None
+    prev_url = None
+    if next_page is not None:
+        next_url = f'{url}?{urlencode({"page": next_page, "page_size": page_size})}'
+    if prev_page is not None:
+        prev_url = f'{url}?{urlencode({"page": prev_page, "page_size": page_size})}'
+
+    return {'results': paths.data, 'count': asset_count, 'next': next_url, 'previous': prev_url}
```

Comment:
i would say that this should not be a data model issue (i would be very hesitant to introduce a directory in our data model). 

it should literally be handled by a good function. this is going to come up in the future for pagination of any sort of search result.  then the question boils down to can an efficient and deterministic function be implemented at a given level of the tree. 

as a thought experiment consider that you want a sort flag to go alongside this pagesort endpoint that's either by type (dir, file) or by date or by name.  (i'm completely fine with not sorting right this minute, but considering it through that lens may show other approaches).

Reply:
@satra, what do you mean by "good function"?

And more generally, I would like to keep a skeptical eye towards the notion that directories and files should not be part of our data model. In fact, those concepts are already part of our model, just not explicitly. And that generally leads to trouble like we have been facing here, where the database doesn't offer enough support around what we actually want to accomplish, instead pushing us to design workarounds that need to be tested, etc. Instead, we could take advantage of the existing tools if we accepted explicitly that our data model does indeed have files and folders mixed into it.

Reply:
> "good function"?

an efficient function with optimizations. and we could do some asset optimizations based on path. e.g. a field could store all levels of the paths of an asset at different levels.  let's say a field simply stored a list of folders, then a count at a given level simply needs to filter all assets with folder length > level

Reply:
I don't think length would be enough. Eventually I think we might end up with extra table for folders tree, and have many to one indexing into it from dandiset assets table, while performing needed changes whenever asset is modified... But versioning support would keep things complicated... until we arrive at git's model ;-)

Reply:
Indeed, we are now talking about tweaking the DB schema to support the operations we think we want to perform--this is the job of the modeling layer. Consider, for example, that a "file" model naturally has a parent directory--that alone would allow us to efficiently query for what to show in the file browser at a particular level. Whether we would denormalize the full chain of parents all the way to the root (in order to optimize displaying the full path in the file browser, or make it easier to count all descendants of a directory, etc.) would be a decision to make on top of designing new models to support this part of the app.

This is something we should probably discuss in the next standup.